doc/docs.md

```diff
@@ -1708,6 +1709,50 @@ fn main(){
 
 # Advanced Topics
 
+## Memory-unsafe code
+
+Sometimes for efficiency you may want to write low-level code that may potentially
+corrupt memory. V supports that, but not by default.
+
+V requires that any potentially memory-unsafe operations are marked intentionally.
+Marking them also indicates to anyone reading the code that there could be memory 
+corruption if there was a mistake. 
+
+Examples of memory-unsafe operations are:
+
+* Pointer arithmetic
+* Pointer indexing
+* Conversion to pointer from an incompatible type
+* Calling a C function
```

Comment:
Calling a C function is not unsafe by default imho. For example calling `fn C.sin(x f64) f64` can not corrupt memory.

Reply:
Yes, but we need to mark C prototypes in V as safe. They should default to unsafe. I suggest a `[trusted]`attribute

Reply:
For now I changed the wording to *Calling certain C functions, e.g. `strlen`, `strncmp`, `free`* - thanks.

Reply:
`strlen` and `strncmp` are not unsafe - they don't modify anything.  `free` is inherently unsafe because it could free memory that is still in use.

Reply:
we already have `[unsafe_fn]` for marking unsafe functions

Reply:
The `[trusted]` attribute and changing the default for C function prototypes may be a good idea indeed. @medvednikov what do you think?

Reply:
@JalonSolov calling C.strlen(0) for example or over a pointer to a block of memory that is not terminated by 0, may be unsafe

Reply:
![strlen_0](https://url4e.com/gyazo/images/33e740dc.png)

Reply:
I don't see how strlen can be unsafe.  You might get a wildly different answer than expected, but the only way for it to be unsafe is if you tried it on a machine with memory-mapped IO... do those even exist any longer?  As for strlen(0)... your code may just abort with an error, but it won't corrupt anything.

Reply:
@JalonSolov
> strlen and strncmp are not unsafe - they don't modify anything.

AFAIK we haven't decided a formal Memory safety definition for V, but Wikipedia says:

> Memory safety is the state of being protected from various software bugs and security vulnerabilities when dealing with memory access, such as buffer overflows and dangling pointers.

(from https://en.wikipedia.org/wiki/Memory_safety)

This page looks a good summary too:
https://medium.com/@arunrajeevan/what-is-memory-safety-in-programming-1e5300fe6c48

I think V should definitely count invalid memory reads as memory-unsafe, they can cause security issues. Another question is whether undefined behaviour is considered `unsafe` - this is likely good to adopt too.

Reply:
@spytheman 
> we already have [unsafe_fn] for marking unsafe functions, but I have not checked if that attribute works for C prototypes

I may be wrong, but I'm not sure it has an effect even for V functions. I plan to make V require `unsafe` blocks to call functions that have that attribute.

I think it's too easy to forget to mark a C prototype as `[unsafe_fn]`, and programmers typically forget about certain types of memory unsafety (myself included in complex code), so we need to always remind them to think about it.

Reply:
@ntrel by that reasoning, you may as well mark any use of C as unsafe, since anything you do with C can be forced to do "unsafe" things.

Reply:
@jalonsolov No, safe V code cannot force any C function to violate memory safety.